File: qt/gen_qlcdnumber.cpp
// +build ignore

#include <QEvent>
#include <QFrame>
#include <QLCDNumber>
#include <QMetaObject>
#include <QObject>
#include <QPaintDevice>
#include <QPaintEvent>
#include <QSize>
#include <QString>
#include <QByteArray>
#include <cstring>
#include <QWidget>
#include <qlcdnumber.h>
#include "gen_qlcdnumber.h"
class MiqtVirtualQLCDNumber : public virtual QLCDNumber {
public:
MiqtVirtualQLCDNumber(QWidget* parent): QLCDNumber(parent) {};
MiqtVirtualQLCDNumber(): QLCDNumber() {};
MiqtVirtualQLCDNumber(uint numDigits): QLCDNumber(numDigits) {};
MiqtVirtualQLCDNumber(uint numDigits, QWidget* parent): QLCDNumber(numDigits, parent) {};

virtual ~MiqtVirtualQLCDNumber() = default;
// cgo.Handle value for overwritten implementation
	intptr_t handle__MetaObject = 0;
// Subclass to allow providing a Go implementation
	virtual const QMetaObject* metaObject() const override {
if (handle__MetaObject == 0) {
return QLCDNumber::metaObject();
}
QMetaObject* callback_return_value = miqt_exec_callback_QLCDNumber_MetaObject(const_cast<MiqtVirtualQLCDNumber*>(this), handle__MetaObject);

		return callback_return_value;
	}
// Wrapper to allow calling protected method
	QMetaObject* virtualbase_MetaObject() const {

		return (QMetaObject*) QLCDNumber::metaObject();

	}
// cgo.Handle value for overwritten implementation
	intptr_t handle__Metacast = 0;
// Subclass to allow providing a Go implementation
	virtual void* qt_metacast(const char* param1) override {
if (handle__Metacast == 0) {
return QLCDNumber::qt_metacast(param1);
}
const char* sigval1 = (const char*) param1;

		void* callback_return_value = miqt_exec_callback_QLCDNumber_Metacast(this, handle__Metacast, sigval1);

		return callback_return_value;
	}
// Wrapper to allow calling protected method
	void* virtualbase_Metacast(const char* param1) {

		return QLCDNumber::qt_metacast(param1);

	}
};
QLCDNumber* QLCDNumber_new(QWidget* parent) {
return new MiqtVirtualQLCDNumber(parent);
}
QLCDNumber* QLCDNumber_new2() {
return new MiqtVirtualQLCDNumber();
}
QLCDNumber* QLCDNumber_new3(unsigned int numDigits) {
return new MiqtVirtualQLCDNumber(static_cast<uint>(numDigits));
}
QLCDNumber* QLCDNumber_new4(unsigned int numDigits, QWidget* parent) {
return new MiqtVirtualQLCDNumber(static_cast<uint>(numDigits), parent);
}
void QLCDNumber_virtbase(QLCDNumber* src
, QFrame** outptr_QFrame
) {
*outptr_QFrame = static_cast<QFrame*>(src);
}
QMetaObject* QLCDNumber_MetaObject(const QLCDNumber* self) {
	return (QMetaObject*) self->metaObject();
}
void* QLCDNumber_Metacast(QLCDNumber* self, const char* param1) {
	return self->qt_metacast(param1);
}
struct miqt_string QLCDNumber_Tr(const char* s) {
	QString _ret = QLCDNumber::tr(s);
	// Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
	QByteArray _b = _ret.toUtf8();
	struct miqt_string _ms;
	_ms.len = _b.length();
	_ms.data = static_cast<char*>(malloc(_ms.len));
	memcpy(_ms.data, _b.data(), _ms.len);
	return _ms;
}
bool QLCDNumber_SmallDecimalPoint(const QLCDNumber* self) {
	return self->smallDecimalPoint();
}
int QLCDNumber_DigitCount(const QLCDNumber* self) {
	return self->digitCount();
}
void QLCDNumber_SetDigitCount(QLCDNumber* self, int nDigits) {
	self->setDigitCount(static_cast<int>(nDigits));
}
bool QLCDNumber_CheckOverflow(const QLCDNumber* self, double num) {
	return self->checkOverflow(static_cast<double>(num));
}
bool QLCDNumber_CheckOverflowWithNum(const QLCDNumber* self, int num) {
	return self->checkOverflow(static_cast<int>(num));
}
Mode QLCDNumber_Mode(const QLCDNumber* self) {
	return self->mode();
}
void QLCDNumber_SetMode(QLCDNumber* self, Mode mode) {
	self->setMode(mode);
}
SegmentStyle QLCDNumber_SegmentStyle(const QLCDNumber* self) {
	return self->segmentStyle();
}
void QLCDNumber_SetSegmentStyle(QLCDNumber* self, SegmentStyle segmentStyle) {
	self->setSegmentStyle(segmentStyle);
}
double QLCDNumber_Value(const QLCDNumber* self) {
	return self->value();
}
int QLCDNumber_IntValue(const QLCDNumber* self) {
	return self->intValue();
}
QSize* QLCDNumber_SizeHint(const QLCDNumber* self) {
	return new QSize(self->sizeHint());
}
void QLCDNumber_Display(QLCDNumber* self, struct miqt_string str) {
	QString str_QString = QString::fromUtf8(str.data, str.len);
	self->display(str_QString);
}
void QLCDNumber_DisplayWithNum(QLCDNumber* self, int num) {
	self->display(static_cast<int>(num));
}
void QLCDNumber_Display2(QLCDNumber* self, double num) {
	self->display(static_cast<double>(num));
}
void QLCDNumber_SetHexMode(QLCDNumber* self) {
	self->setHexMode();
}
void QLCDNumber_SetDecMode(QLCDNumber* self) {
	self->setDecMode();
}
void QLCDNumber_SetOctMode(QLCDNumber* self) {
	self->setOctMode();
}
void QLCDNumber_SetBinMode(QLCDNumber* self) {
	self->setBinMode();
}
void QLCDNumber_SetSmallDecimalPoint(QLCDNumber* self, bool smallDecimalPoint) {
	self->setSmallDecimalPoint(smallDecimalPoint);
}
void QLCDNumber_Overflow(QLCDNumber* self) {
	self->overflow();
}
void QLCDNumber_connect_Overflow(QLCDNumber* self, intptr_t slot) {
	MiqtVirtualQLCDNumber::connect(self, static_cast<void (QLCDNumber::*)()>(&QLCDNumber::overflow), self, [=]() {
		miqt_exec_callback_QLCDNumber_Overflow(slot);
	});
}
struct miqt_string QLCDNumber_Tr2(const char* s, const char* c) {
	QString _ret = QLCDNumber::tr(s, c);
	// Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
	QByteArray _b = _ret.toUtf8();
	struct miqt_string _ms;
	_ms.len = _b.length();
	_ms.data = static_cast<char*>(malloc(_ms.len));
	memcpy(_ms.data, _b.data(), _ms.len);
	return _ms;
}
struct miqt_string QLCDNumber_Tr3(const char* s, const char* c, int n) {
	QString _ret = QLCDNumber::tr(s, c, static_cast<int>(n));
	// Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
	QByteArray _b = _ret.toUtf8();
	struct miqt_string _ms;
	_ms.len = _b.length();
	_ms.data = static_cast<char*>(malloc(_ms.len));
	memcpy(_ms.data, _b.data(), _ms.len);
	return _ms;
}
void QLCDNumber_override_virtual_MetaObject(void* self, intptr_t slot) {
	dynamic_cast<MiqtVirtualQLCDNumber*>( (QLCDNumber*)(self) )->handle__MetaObject = slot;
}
QMetaObject* QLCDNumber_virtualbase_MetaObject(const void* self) {
	return ( (const MiqtVirtualQLCDNumber*)(self) )->virtualbase_MetaObject();
}
void QLCDNumber_override_virtual_Metacast(void* self, intptr_t slot) {
	dynamic_cast<MiqtVirtualQLCDNumber*>( (QLCDNumber*)(self) )->handle__Metacast = slot;
}
void* QLCDNumber_virtualbase_Metacast(void* self, const char* param1) {
	return ( (MiqtVirtualQLCDNumber*)(self) )->virtualbase_Metacast(param1);
}
void QLCDNumber_Delete(QLCDNumber* self, bool isSubclass) {
	if (isSubclass) {
		delete dynamic_cast<MiqtVirtualQLCDNumber*>( self );
	} else {
		delete self;
	}
}

File: qt/gen_qlcdnumber.go
package qt

import (
	"unsafe"
)

type QLCDNumber__Mode int

const (
	QLCDNumber__Hex QLCDNumber__Mode = 0
	QLCDNumber__Dec QLCDNumber__Mode = 1
	QLCDNumber__Oct QLCDNumber__Mode = 2
	QLCDNumber__Bin QLCDNumber__Mode = 3
)

type QLCDNumber__SegmentStyle int

const (
	QLCDNumber__Outline QLCDNumber__SegmentStyle = 0
	QLCDNumber__Filled  QLCDNumber__SegmentStyle = 1
	QLCDNumber__Flat    QLCDNumber__SegmentStyle = 2
)

type QLCDNumber struct {
	h          uintptr
	isSubclass bool
}

// NewQLCDNumber constructs a new QLCDNumber object.
func NewQLCDNumber(parent *QWidget) *QLCDNumber {
	g := newQLCDNumber(QLCDNumber_new(parent.cPointer()))
	g.isSubclass = true
	return g
}

// NewQLCDNumber2 constructs a new QLCDNumber object.
func NewQLCDNumber2() *QLCDNumber {
	g := newQLCDNumber(QLCDNumber_new2())
	g.isSubclass = true
	return g
}

// NewQLCDNumber3 constructs a new QLCDNumber object.
func NewQLCDNumber3(numDigits uint) *QLCDNumber {
	g := newQLCDNumber(QLCDNumber_new3((uint)(numDigits)))
	g.isSubclass = true
	return g
}

// NewQLCDNumber4 constructs a new QLCDNumber object.
func NewQLCDNumber4(numDigits uint, parent *QWidget) *QLCDNumber {
	g := newQLCDNumber(QLCDNumber_new4((uint)(numDigits), parent.cPointer()))
	g.isSubclass = true
	return g
}

func (this *QLCDNumber) MetaObject() *QMetaObject {
	return newQMetaObject(QLCDNumber_MetaObject(this.h))
}

func (this *QLCDNumber) Metacast(param1 string) unsafe.Pointer {
	param1_Cstring := CString(param1)
	defer free(unsafe.Pointer(param1_Cstring))
	return (unsafe.Pointer)(QLCDNumber_Metacast(this.h, param1_Cstring))
}

func QLCDNumber_Tr(s string) string {
	s_Cstring := CString(s)
	defer free(unsafe.Pointer(s_Cstring))
	var _ms struct_miqt_string = QLCDNumber_Tr(s_Cstring)
	_ret := GoStringN(_ms.data, int(int64(_ms.len)))
	free(unsafe.Pointer(_ms.data))
	return _ret
}

func (this *QLCDNumber) SmallDecimalPoint() bool {
	return (bool)(QLCDNumber_SmallDecimalPoint(this.h))
}

func (this *QLCDNumber) DigitCount() int {
	return (int)(QLCDNumber_DigitCount(this.h))
}

func (this *QLCDNumber) SetDigitCount(nDigits int) {
	QLCDNumber_SetDigitCount(this.h, (int)(nDigits))
}

func (this *QLCDNumber) CheckOverflow(num float64) bool {
	return (bool)(QLCDNumber_CheckOverflow(this.h, (double)(num)))
}

func (this *QLCDNumber) CheckOverflowWithNum(num int) bool {
	return (bool)(QLCDNumber_CheckOverflowWithNum(this.h, (int)(num)))
}

func (this *QLCDNumber) Mode() Mode {
	xxxxxxxxx
}

func (this *QLCDNumber) SetMode(mode Mode) {
	QLCDNumber_SetMode(this.h, mode)
}

func (this *QLCDNumber) SegmentStyle() SegmentStyle {
	xxxxxxxxx
}

func (this *QLCDNumber) SetSegmentStyle(segmentStyle SegmentStyle) {
	QLCDNumber_SetSegmentStyle(this.h, segmentStyle)
}

func (this *QLCDNumber) Value() float64 {
	return (float64)(QLCDNumber_Value(this.h))
}

func (this *QLCDNumber) IntValue() int {
	return (int)(QLCDNumber_IntValue(this.h))
}

func (this *QLCDNumber) SizeHint() *QSize {
	_goptr := newQSize(QLCDNumber_SizeHint(this.h))
	_goptr.GoGC() // Qt uses pass-by-value semantics for this type. Mimic with finalizer
	return _goptr
}

func (this *QLCDNumber) Display(str string) {
	str_ms := struct_miqt_string{}
	str_ms.data = CString(str)
	str_ms.len = size_t(len(str))
	defer free(unsafe.Pointer(str_ms.data))
	QLCDNumber_Display(this.h, str_ms)
}

func (this *QLCDNumber) DisplayWithNum(num int) {
	QLCDNumber_DisplayWithNum(this.h, (int)(num))
}

func (this *QLCDNumber) Display2(num float64) {
	QLCDNumber_Display2(this.h, (double)(num))
}

func (this *QLCDNumber) SetHexMode() {
	QLCDNumber_SetHexMode(this.h)
}

func (this *QLCDNumber) SetDecMode() {
	QLCDNumber_SetDecMode(this.h)
}

func (this *QLCDNumber) SetOctMode() {
	QLCDNumber_SetOctMode(this.h)
}

func (this *QLCDNumber) SetBinMode() {
	QLCDNumber_SetBinMode(this.h)
}

func (this *QLCDNumber) SetSmallDecimalPoint(smallDecimalPoint bool) {
	QLCDNumber_SetSmallDecimalPoint(this.h, (bool)(smallDecimalPoint))
}

func (this *QLCDNumber) Overflow() {
	QLCDNumber_Overflow(this.h)
}

func (this *QLCDNumber) OnOverflow(slot func()) {
	QLCDNumber_connect_Overflow(this.h, intptr_t(cgo.NewHandle(slot)))
}

//export miqt_exec_callback_QLCDNumber_Overflow
func miqt_exec_callback_QLCDNumber_Overflow(cb intptr_t) {
	gofunc, ok := cgo.Handle(cb).Value().(func())
	if !ok {
		panic("miqt: callback of non-callback type (heap corruption?)")
	}

	gofunc()
}

func QLCDNumber_Tr2(s string, c string) string {
	s_Cstring := CString(s)
	defer free(unsafe.Pointer(s_Cstring))
	c_Cstring := CString(c)
	defer free(unsafe.Pointer(c_Cstring))
	var _ms struct_miqt_string = QLCDNumber_Tr2(s_Cstring, c_Cstring)
	_ret := GoStringN(_ms.data, int(int64(_ms.len)))
	free(unsafe.Pointer(_ms.data))
	return _ret
}

func QLCDNumber_Tr3(s string, c string, n int) string {
	s_Cstring := CString(s)
	defer free(unsafe.Pointer(s_Cstring))
	c_Cstring := CString(c)
	defer free(unsafe.Pointer(c_Cstring))
	var _ms struct_miqt_string = QLCDNumber_Tr3(s_Cstring, c_Cstring, (int)(n))
	_ret := GoStringN(_ms.data, int(int64(_ms.len)))
	free(unsafe.Pointer(_ms.data))
	return _ret
}

func (this *QLCDNumber) callVirtualBase_MetaObject() *QMetaObject {
	return newQMetaObject(QLCDNumber_virtualbase_MetaObject(unsafe.Pointer(this.h)))
}

func (this *QLCDNumber) OnMetaObject(slot func(super func() *QMetaObject) *QMetaObject) {
	if !this.isSubclass {
		panic("miqt: can only override virtual methods for directly constructed types")
	}
	QLCDNumber_override_virtual_MetaObject(unsafe.Pointer(this.h), intptr_t(cgo.NewHandle(slot)))
}

//export miqt_exec_callback_QLCDNumber_MetaObject
func miqt_exec_callback_QLCDNumber_MetaObject(self QLCDNumber, cb intptr_t) *QMetaObject {
	gofunc, ok := cgo.Handle(cb).Value().(func(super func() *QMetaObject) *QMetaObject)
	if !ok {
		panic("miqt: callback of non-callback type (heap corruption?)")
	}

	virtualReturn := gofunc((&QLCDNumber{h: self}).callVirtualBase_MetaObject)

	return virtualReturn.cPointer()
}

func (this *QLCDNumber) callVirtualBase_Metacast(param1 string) unsafe.Pointer {
	param1_Cstring := CString(param1)
	defer free(unsafe.Pointer(param1_Cstring))

	return (unsafe.Pointer)(QLCDNumber_virtualbase_Metacast(unsafe.Pointer(this.h), param1_Cstring))
}

func (this *QLCDNumber) OnMetacast(slot func(super func(param1 string) unsafe.Pointer, param1 string) unsafe.Pointer) {
	if !this.isSubclass {
		panic("miqt: can only override virtual methods for directly constructed types")
	}
	QLCDNumber_override_virtual_Metacast(unsafe.Pointer(this.h), intptr_t(cgo.NewHandle(slot)))
}

//export miqt_exec_callback_QLCDNumber_Metacast
func miqt_exec_callback_QLCDNumber_Metacast(self QLCDNumber, cb intptr_t, param1 *const_char) unsafe.Pointer {
	gofunc, ok := cgo.Handle(cb).Value().(func(super func(param1 string) unsafe.Pointer, param1 string) unsafe.Pointer)
	if !ok {
		panic("miqt: callback of non-callback type (heap corruption?)")
	}
	// Convert all CABI parameters to Go parameters
	param1_ret := param1
	slotval1 := GoString(param1_ret)
	virtualReturn := gofunc((&QLCDNumber{h: self}).callVirtualBase_Metacast, slotval1)

	return virtualReturn
}
